src/s3.rs

```diff
@@ -12425,6 +12492,17 @@ impl<P, D> S3Client<P, D> where P: ProvideAwsCredentials, D: DispatchSignedReque
             None => host,
         }
     }
+
+    fn is_dns_compatible(&self, bucket: &BucketName) -> bool {
+        self.is_valid_subdomain(&bucket) && !bucket.contains(".")
+    }
+
+    fn is_valid_subdomain(&self, bucket: &BucketName) -> bool {
+        let subdomain_re: Regex = Regex::new(r"^[a-z0-9][a-z0-9.-]+[a-z0-9]$").unwrap();
```

Comment:
We could use `lazy_static!` to avoid building these many times.  See [here](https://github.com/rusoto/rusoto/blob/master/codegen/src/generator/rest_json.rs#L128) for an example.